python/ucxx/_lib/libucxx.pyx

```diff
@@ -974,6 +984,34 @@ cdef class UCXEndpoint():
             handle = self._endpoint.get().getHandle()
 
         return int(<uintptr_t>handle)
+    
+    @property
+    def ucxx_ptr(self):
+        cdef Endpoint* endpoint
+
+        with nogil:
+            endpoint = self._endpoint.get()
+
+        return int(<uintptr_t>endpoint)
+        # return int(<uintptr_t><void*>&self._endpoint)
+            
+    @property
+    def worker_handle(self):
+        cdef ucp_worker_h handle
+
+        with nogil:
+            handle = self._endpoint.get().getWorker().get().getHandle()
+
+        return int(<uintptr_t>handle)
+    
+    @property
+    def ucxx_worker_ptr(self):
```

Comment:
I'm not entirely certain either, I don't know enough of how Cython actually handles all cleanups in detail so I went for the "safer" route. You're probably right that constructing the worker via a `@property` is probably find, and maybe even holding onto it though. I guess we could in theory still do:

```
...
worker = ucx_api.UCXWorker(ctx)
ep = ucx_api.UCXEndpoint.create(worker, ...)

worker_copy = ep.worker  # `UCXEndpoint.worker` is now a `@property` that constructs a `UCXWorker` from the underlying `std::shared_ptr<ucxx::Worker>`
worker_handle = worker_copy.ucxx_ptr
```

When the above executes, eventually all `worker`, `ep` and `worker_copy` would go out-of-scope and call their destructors which will decrement the refcount of the `std::shared_ptr<Worker>`, and when it goes to zero then its C++ destructor would be called. The problem would only exist if the user would keep `worker_copy` somewhere forever, but that could be just as problematic as keeping `worker` itself, so perhaps we have nothing to worry with doing what you said.

Do you think that feels like the right way to think about this?

Reply:
I _think_ so, but let us not do that here, because I am not really sure of myself and need to think through the consequences in a bit more detail. I think you are right that if the user holds on to the worker forever it doesn't matter if it is the "copy" or the original one. The issue I am (potentially) worried about is setting up a ref-cycle internally such that the python ucxx bindings preclude collection by refcounts at the python level going to zero.

Reply:
I don't. Please go ahead!